[dagster-dbt] add dbt cloud logs to stdout after command runs

cnolanminich opened the pull request:
## Summary & Motivation

Our dbt core integration will show logs to stdout for the dbt command that was run, but we don't do that for dbt cloud. Per the dbt cloud docs, it does not have a streaming log API endpoint, but you can get the logs after the dbt commands complete.


Example output:

<img width="1290" height="858" alt="image" src="https://github.com/user-attachments/assets/a0ffef97-a353-48ec-8156-0b68ec435927" />


## How I Tested These Changes

BK + pytest (locally) + our dbt cloud test environment


## Changelog

* [dagster-dbt] Add logs for dbt cloud to the stdout after the run completes in dbt cloud

cnolanminich commented:
All the BK failures are due to  identical failure across several versions of dbt-core (unrelated to my change above):


###  Failing Tests (same in all 4 logs):
  1. test_column_lineage[False-snowflake---select fqn:*]
  2. test_column_lineage[False-trino---select fqn:*]


### The Specific Issue:

  # What the test gets:
  AssetKey(['duplicate_column_dep_orders']): None

  # What the test expects:
  AssetKey(['duplicate_column_dep_orders']): TableColumnLineage(
      deps_by_column={'amount_2x': [TableColumnDep(asset_key=AssetKey(['orders']), column_name='amount')]}
  )

  The column lineage metadata for the duplicate_column_dep_orders dbt model is not being extracted.

 ### Failure Conditions:

  1. ✅ Parameter: use_async_fetch_column_schema=False
    - Tests with True PASS ✅
  2. ✅ SQL Dialects: Only snowflake and trino
    - bigquery, databricks, duckdb PASS ✅

  3. ✅ Selection: Only --select fqn:* (all models at once)
    - Individual model selections PASS ✅

OwenKephart commented on python_modules/libraries/dagster-dbt/dagster_dbt/cloud_v2/cli_invocation.py:
```diff
@@ -54,6 +54,12 @@ def wait(
+        # Write dbt Cloud run logs to stdout
+        logs = self.run_handler.get_run_logs()
+        if logs:
+            print(logs, flush=True)  # noqa: T201
```
should use `sys.stdout.write()` instead of print (that'll get rid of that noqa for you as well)

OwenKephart commented on python_modules/libraries/dagster-dbt/dagster_dbt/cloud_v2/client.py:
```diff
@@ -355,18 +364,46 @@ def list_run_artifacts(
+        url = f"{self.api_v2_url}/runs/{run_id}/artifacts/{path}"
+
+        num_retries = 0
+        while True:
```
why is this change necessary? seems like it's basically the same logic as the existing _make_request method already has, so if there's something lacking about that method i'd prefer if we just updated that instead of replicating it here

cnolanminich replied:
not necessary -- wasn't sure the best way to handle that `_make_requests()` now might return text or JSON while keeping the return type as a dictionary. Not 100% happy with the current solution I just put in (trying to parse the JSON and then adding "text"), but it looks like the http response header from dbt cloud doesn't always include an "application/json" for json so I'm not sure if there's a better way

cnolanminich commented:
@OwenKephart  made your suggested edits, modified the return of `_make_requests()` to be a generic `requests.Response` and then parsing as JSON or text depending on the downstream use.

I also found that sometimes the logs aren't ready for short-lived dbt run commands, so I added retries when a step does not have logs yet (configured with 3 retries but it seems like it's a 1-3 second thing).
